### PR TITLE
feat: Fetch contacts/recipients for messages

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Add the ability to fetch recipients (clients of users) for a given conversation

### Solutions

No UseCase implemented, as this is _probably_ not needed by users of this library directly, but only used by other UseCases.

### Testing

Covered mappers with tests.
Repository not yet as they are subject to change, because persistence is not in place yet.

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt [hunk 1]
```diff
@@ -0,0 +1,91 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.android.core.exception.Failure
+import com.wire.android.core.functional.Either
+import com.wire.android.core.functional.suspending
+import com.wire.android.feature.contact.DetailedContact
+import com.wire.android.feature.conversation.content.MessageRepository
+import com.wire.android.feature.conversation.data.ConversationRepository
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt:8]: We are pulling another `ConversationRepository`? Where does this come from?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        conversationRepository.detailedConversationMembers(conversationId).flatMap { detailedContacts ->
+            createSessionForMissingClientsIfNeeded(senderUserId, detailedContacts).map {
+                detailedContacts
+            }
+        }
+    }
+
+    private suspend fun createSessionForMissingClientsIfNeeded(
+        senderUserId: String,
+        detailedContacts: List<DetailedContact>
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt:40]: I find it confusing to use both contact and user when referring to a "user".

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                        userPreKeyInfo.userId,
+                        clientPreKeyInfo.clientId,
+                        clientPreKeyInfo.preKey
+                    )
+                }
+            }
+        }
+    }
+
+    private suspend fun missingContactClients(detailedContacts: List<DetailedContact>, senderUserId: String) =
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt:67]:
```suggestion
    private suspend fun missingClients(detailedContacts: List<DetailedContact>, senderUserId: String) =
```

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt [hunk 4]
```diff
@@ -1,19 +1,70 @@
...
             TODO("Error handling. Repository layer, a good place to use Either<Failure,Success> ?")
         }
         return conversationsResponse.value.conversations.map(conversationMapper::fromApiModel)
     }
 
+    override suspend fun getConversationDetails(conversationId: ConversationId): Either<CoreFailure, Conversation> {
+        val conversationResponse = conversationApi.fetchConversationDetails(conversationId)
+
+        if (!conversationResponse.isSuccessful()) {
+            return Either.Left(CoreFailure.ServerMiscommunication)
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:40]: We can treat that, yes.
I'll add it here.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt (hunk 1 above)
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt:8]: This whole class was actually not supposed to be in this PR. This is on higher-level on top of fetching pre-keys and recipients of a conversation.

The pre-keys part is on its way, and then I'll add another PR with this.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt (hunk 2 above)
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/OutgoingMessageRecipientsRetriever.kt:40]: As I mentioned before, this `OutgoingMessageRecipientsRetriever` actually shouldn't be in this PR yet, I let it leak by accident.

I want to review these names and alter some stuff inside of it when the time to submit it comes.

My suggestion is something along these lines:

|Term | Description|
|----|---|
| `User` | Any user, generic. Can be "self" user. _i.e._ the user that is currently operating this client, or any other when in context of searching for connections and vieweing other profiles, for example. |
| `Contact` | Another user that this self-user has a conversation with. Used in context of conversations. Maybe can be replaced by `ConversationMember`? |
| `Recipient` | Grouping of `Contact`/`ConversationMember` + its Client IDs. Used in context of creating outgoing messages / sending messages, encrypting, creating an envelope in general |


This `DetailedContact` would be replaced with `Recipient` in this case. The `senderUserId` won't be needed at all for this operation, as it is injected via the `sessionScope`.

**Review by MohamadJaara — CHANGES_REQUESTED** [review]
great work, just some small changes :wrench:

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt [hunk 5]
```diff
@@ -1,19 +1,75 @@
...
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.functional.suspending
+import com.wire.kalium.network.api.ConversationId
 import com.wire.kalium.network.api.conversation.ConversationApi
+import com.wire.kalium.network.api.user.client.ClientApi
 import com.wire.kalium.network.utils.isSuccessful
 
-class ConversationRepository(
+interface ConversationRepository {
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:15]: I was thinking of changing later, so the UseCases won't notice the change. Wanna get a conversation? Ok, the repository will keep getting it for the UseCase, but the internals will change.

File: logic/build.gradle.kts [hunk 6]
```diff
@@ -48,6 +48,7 @@ kotlin {
         val commonTest by getting {
             dependencies {
                 implementation(kotlin("test"))
+                implementation(Dependencies.Test.mockative)
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/build.gradle.kts]: Done

**Review by MohamadJaara — APPROVED** [review]
well done!

**Pull request merged**